BlogTemplate/Pages/Edit.cshtml.cs

```diff
@@ -26,72 +24,70 @@ public EditModel(BlogDataStore dataStore, SlugGenerator slugGenerator, ExcerptGe
         }
 
         [BindProperty]
-        public Post newPost { get; set; }
-
-        public Post oldPost { get; set; }
+        public EditedPostModel EditedPost { get; set; }
 
         public void OnGet([FromRoute] int id)
         {
-            newPost = oldPost = _dataStore.GetPost(id);
+            Post post = _dataStore.GetPost(id);
+
+            EditedPost = new EditedPostModel
+            {
+                Title = post.Title,
+                Body = post.Body,
+                Excerpt = post.Excerpt,
+            };
 
-            if (oldPost == null)
+            if (post == null)
             {
                 RedirectToPage("/Index");
             }
         }
 
-
         [ValidateAntiForgeryToken]
-        public IActionResult OnPostPublish([FromRoute] int id)
+        public IActionResult OnPostPublish([FromRoute] int id, [FromForm] bool updateSlug)
         {
-            oldPost = _dataStore.GetPost(id);
-            newPost.IsPublic = true;
-            UpdatePost(id);
-            return Redirect($"/Post/{id}/{newPost.Slug}");
+            Post post = _dataStore.GetPost(id);
+            if (ModelState.IsValid)
+            {
+                bool wasPublic = post.IsPublic;
+                post.IsPublic = true;
+                UpdatePost(post, updateSlug, wasPublic);
+            }
+            return Redirect($"/Post/{id}/{post.Slug}");
         }
 
         [ValidateAntiForgeryToken]
-        public IActionResult OnPostSaveDraft([FromRoute] int id)
-        {
-            oldPost = _dataStore.GetPost(id);
-            newPost.IsPublic = false;
-            UpdatePost(id);
-            return Redirect("/Index");
-        }
-
-        private void UpdatePost(int id)
+        public IActionResult OnPostSaveDraft([FromRoute] int id, [FromForm] bool updateSlug)
         {
-            newPost.Id = id;
-            oldPost = _dataStore.GetPost(id);
-
-            if (oldPost.PubDate.Equals(default(DateTimeOffset)))

```

Comment:
Since this method saves the draft, it shouldn't deal with setting the PubDate for the post. The PubDate for a post is set only once -- when the post is published for the first time. So, if the draft has not yet been published, we don't yet know what its PubDate is.

I think we should keep this logic but move it to the OnPostPublish method because we would only set the PubDate when a post is being published.